File: frontend/src/api/costs.ts
import { api } from './index';

export interface CostOptimizationData {
  costs: {
    compute: number;
    storage: number;
    network: number;
  };
  total_cost: number;
  potential_savings: number;
  savings_percentage: number;
  recommendations: Array<{
    id: string;
    title: string;
    description: string;
    impact: number;
  }>;
}

export const fetchCostOptimization = async (): Promise<CostOptimizationData> => {
  try {
    const response = await api.get('/costs/optimization');
    return response.data;
  } catch (error) {
    console.warn('Failed to fetch cost optimization data, using mock data:', error);
    // Return mock data for development
    return {
      costs: {
        compute: 1200,
        storage: 800,
        network: 400
      },
      total_cost: 2400,
      potential_savings: 600,
      savings_percentage: 25,
      recommendations: [
        {
          id: '1',
          title: 'Optimize compute resources',
          description: 'Consider right-sizing underutilized instances',
          impact: 300
        },
        {
          id: '2',
          title: 'Storage optimization',
          description: 'Move infrequently accessed data to lower-cost storage tiers',
          impact: 200
        }
      ]
    };
  }
};

File: backend/app/routers/costs.py
from fastapi import APIRouter, HTTPException
from typing import List
from pydantic import BaseModel
from datetime import datetime

router = APIRouter()

class Recommendation(BaseModel):
    id: str
    title: str
    description: str
    impact: float

class CostBreakdown(BaseModel):
    compute: float
    storage: float
    network: float

class CostOptimizationResponse(BaseModel):
    costs: CostBreakdown
    total_cost: float
    potential_savings: float
    savings_percentage: float
    recommendations: List[Recommendation]

@router.get("/costs/optimization", response_model=CostOptimizationResponse)
async def get_cost_optimization():
    """
    Get cost optimization data including current costs, potential savings,
    and recommendations for cost reduction.
    """
    try:
        # In a real implementation, this would fetch data from your cloud provider's API
        # or your cost management database
        return {
            "costs": {
                "compute": 1200.00,
                "storage": 800.00,
                "network": 400.00
            },
            "total_cost": 2400.00,
            "potential_savings": 600.00,
            "savings_percentage": 25.0,
            "recommendations": [
                {
                    "id": "1",
                    "title": "Optimize compute resources",
                    "description": "Consider right-sizing underutilized instances based on CPU and memory usage patterns",
                    "impact": 300.00
                },
                {
                    "id": "2",
                    "title": "Storage optimization",
                    "description": "Move infrequently accessed data to lower-cost storage tiers",
                    "impact": 200.00
                },
                {
                    "id": "3",
                    "title": "Network cost reduction",
                    "description": "Optimize data transfer patterns and use caching where possible",
                    "impact": 100.00
                }
            ]
        }
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to retrieve cost optimization data: {str(e)}"
        )
